Upgrading Mirage Packages

patricoferris opened the pull request:
A few remarks about this PR:

 - A few functor applications are gone \o/ (thanks @hannesm et al. for their work on defunctorising parts of the Mirage stack in favour of dune virtual libraries).
 - The TCP stack is becoming more agnostic to IPV4 and IPV6, this PR needs a little work in terms of renaming modules and variables to reflect those changes.
 - These changes allow VPNkit to be built with OCaml 5+ compilers :) (the original motivation behind the PR).

This PR will depend on updating hvsock (https://github.com/mirage/ocaml-hvsock/pull/69) and ocaml-9p (https://github.com/mirage/ocaml-9p/pull/146).

I was also struggling to build on nixos, but I'll work on that.

patricoferris committed:
Update Mirage packages

This also allows VPNkit to be build with OCaml 5+

Signed-off-by: Patrick Ferris <patrick@sirref.org>

hannesm commented on src/dns_mirage/dns_resolver_mirage.mli:
```diff
@@ -52,4 +52,4 @@ type static_dns =
 module Static : S with type stack = static_dns
 
-module Make(Time:Mirage_time.S)(S:Tcpip.Stack.V4) : S with type stack = S.t
+module Make(Time:Mirage_time.S)(S:Tcpip.Stack.V4V6) : S with type stack = S.t
```
FWIW the Time parameter can be abandoned and all calls replaced by Mirage_sleep.ns... no need to depend on mirage-time (neither on mirage-clock) anymore.

patricoferris replied:
Indeed! Done in https://github.com/moby/vpnkit/pull/646/commits/1a4f62ae0243413d861af97854c0f55ea91fe0f9

hannesm commented on src/hostnet/hostnet_icmp.ml:
```diff
@@ -145,9 +145,9 @@ module Make
-            Icmpv4_wire.set_icmpv4_id raw id';
-            Icmpv4_wire.set_icmpv4_csum raw 0;
-            Icmpv4_wire.set_icmpv4_csum raw (Tcpip_checksum.ones_complement raw);
+            Cstruct.BE.set_uint16 raw 5 id'; (* TODO: Upstream *)
```
nice catch, please upstream the missing bindings (from when I removed ppx_cstruct)

hannesm commented on src/hostnet_test/dune:
```diff
@@ -2,5 +2,5 @@
  (libraries hostnet cmdliner alcotest logs.fmt protocol-9p mirage_dns
    mirage-clock-unix charrua-client.mirage forwarder sha
-   mirage-random-stdlib arp.mirage)
+   mirage-random-stdlib arp.mirage tcpip.ipv6 mirage-crypto-rng.unix)
```
the mirage-random-stdlib is to be avoided (deprecated) and shouldn't be needed. Just use Mirage_crypto_rng directly (and Mirage_crypto_rng_unix.use_default ()).

patricoferris committed:
Remove all Mirage_time.S and Mirage_clock.S

Signed-off-by: Patrick Ferris <patrick@sirref.org>

hannesm commented on vpnkit.opam:
```diff
@@ -64,7 +63,6 @@ depends: [
   "mirage-clock" {>= "4.0.0"}
   "mirage-clock-unix" {>= "4.0.0"}
   "mirage-random" {>= "3.0.0"}
```
```suggestion
  "mirage-sleep" {>= "4.0.0"}
  "mirage-mtime" {>= "5.0.0"}
```

patricoferris committed:
Update opam file

Signed-off-by: Patrick Ferris <patrick@sirref.org>

patricoferris committed:
Remove mirage-clock-unix

Signed-off-by: Patrick Ferris <patrick@sirref.org>

patricoferris committed:
Upper-bound cohttp

djs55 committed:
opam: remove mirage-entropy and relax upper bounds

Signed-off-by: David Scott <dave@recoil.org>

djs55 committed:
update Dockerfile for opam 2.2 and ocaml 5.3

In particular io-page-unix needs recent opam

Signed-off-by: David Scott <dave@recoil.org>

djs55 committed:
dune: remove result (built-in)

Signed-off-by: David Scott <dave@recoil.org>

djs55 committed:
dune: remove mirage-profile and mirage-clock

Signed-off-by: David Scott <dave@recoil.org>

djs55 commented:
I've pushed a few more patches, now the Dockerfile should build with OCaml 5.3 and opam 2.2

djs55 commented:
I'll rebase against master since there are now some integration tests running in CI

djs55 committed:
fix issue with bridge IPvs and duplicate MACs

There is some unexpected hidden state somewhere which is causing
with_stack called in test case with the same UUID as a previous case,
to discover the same IP binding ("Reconnecting ...") Perhaps the DHCP
server remembers the MAC?

Signed-off-by: David Scott <dave@recoil.org>

djs55 committed:
update licenses

Signed-off-by: David Scott <dave@recoil.org>

hannesm commented:
Dear @djs55, impressive work. For the mirage-time/fake-time, I don't think dune variants work across packages (they used to, but there was some bad things happening).

But we have a mirage-mtime.mock available (see https://github.com/mirage/mirage-mtime/tree/main/mock), used i. mirage-tcpip tests https://github.com/mirage/mirage-tcpip/tree/main/test/mock-clock

I'm on vacation without my laptop for the next weeks, so please excusive my brevity, and if I'm missing something (hard to grasp big changes on a tiny mobile screen).

hannesm commented on src/hostnet/dune:
```diff
@@ -7,7 +7,7 @@
    mirage-vnetif uuidm cohttp-lwt mirage-channel ezjsonm
-   duration mirage-time mirage-clock
-   mirage-random tcpip.checksum forwarder cstructs sha)
+   duration mirage-time
```
I guess you can drop mirage-time now here as well.

hannesm committed:
some minor tweak, reuse the mirage-mtime.mock implementation

* some minor tweak, reuse the mirage-mtime.mock implementation
* opam: update bounds (requires opam releases being merged)

hannesm commented:
There are two ocaml-ci errors:

```
(cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w @1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-40 -strict-sequence -strict-formats -short-paths -keep-locs -warn-error +A-67-69 -g -bin-annot -I src/dns_forward_test/.test.eobjs/byte -I /home/opam/.opam/4.14/lib/alcotest -I /home/opam/.opam/4.14/lib/alcotest/engine -I /home/opam/.opam/4.14/lib/alcotest/stdlib_ext -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/bytes -I /home/opam/.opam/4.14/lib/cmdliner -I /home/opam/.opam/4.14/lib/cstruct -I /home/opam/.opam/4.14/lib/cstruct-lwt -I /home/opam/.opam/4.14/lib/domain-name -I /home/opam/.opam/4.14/lib/duration -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/fmt/cli -I /home/opam/.opam/4.14/lib/fmt/tty -I /home/opam/.opam/4.14/lib/ipaddr -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/logs/fmt -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/lwt-dllist -I /home/opam/.opam/4.14/lib/lwt/unix -I /home/opam/.opam/4.14/lib/macaddr -I /home/opam/.opam/4.14/lib/mirage-channel -I /home/opam/.opam/4.14/lib/mirage-flow -I /home/opam/.opam/4.14/lib/mirage-flow-combinators -I /home/opam/.opam/4.14/lib/mirage-mtime/unix -I /home/opam/.opam/4.14/lib/mirage-sleep/unix -I /home/opam/.opam/4.14/lib/mtime -I /home/opam/.opam/4.14/lib/mtime/clock -I /home/opam/.opam/4.14/lib/ocaml/threads -I /home/opam/.opam/4.14/lib/ocplib-endian -I /home/opam/.opam/4.14/lib/ocplib-endian/bigstring -I /home/opam/.opam/4.14/lib/parsexp -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/re/str -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stdlib-shims -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri/services -I /home/opam/.opam/4.14/lib/uutf -I src/dns/.dns.objs/byte -I src/dns_forward/.dns_forward.objs/byte -I src/dns_forward_lwt_unix/.dns_forward_lwt_unix.objs/byte -no-alias-deps -opaque -open Dune__exe -o src/dns_forward_test/.test.eobjs/byte/dune__exe__Test_fake.cmo -c -impl src/dns_forward_test/test_fake.ml)
File "src/dns_forward_test/test_fake.ml", line 67, characters 6-31:
67 |       Mirage_mtime_set.tick_for Duration.(of_sec 1);
^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module Mirage_mtime_set
```

and 

```
(cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w @1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-40 -strict-sequence -strict-formats -short-paths -keep-locs -warn-error +A-67-69 -g -bin-annot -I src/dns_forward_test/.test_fake.eobjs/byte -I /home/opam/.opam/4.14/lib/alcotest -I /home/opam/.opam/4.14/lib/alcotest/engine -I /home/opam/.opam/4.14/lib/alcotest/stdlib_ext -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/bytes -I /home/opam/.opam/4.14/lib/cmdliner -I /home/opam/.opam/4.14/lib/cstruct -I /home/opam/.opam/4.14/lib/cstruct-lwt -I /home/opam/.opam/4.14/lib/domain-name -I /home/opam/.opam/4.14/lib/duration -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/fmt/cli -I /home/opam/.opam/4.14/lib/fmt/tty -I /home/opam/.opam/4.14/lib/ipaddr -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/logs/fmt -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/lwt-dllist -I /home/opam/.opam/4.14/lib/lwt/unix -I /home/opam/.opam/4.14/lib/macaddr -I /home/opam/.opam/4.14/lib/mirage-channel -I /home/opam/.opam/4.14/lib/mirage-flow -I /home/opam/.opam/4.14/lib/mirage-flow-combinators -I /home/opam/.opam/4.14/lib/mirage-mtime/mock -I /home/opam/.opam/4.14/lib/mirage-mtime/set -I /home/opam/.opam/4.14/lib/mirage-sleep/unix -I /home/opam/.opam/4.14/lib/ocaml/threads -I /home/opam/.opam/4.14/lib/ocplib-endian -I /home/opam/.opam/4.14/lib/ocplib-endian/bigstring -I /home/opam/.opam/4.14/lib/parsexp -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/re/str -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stdlib-shims -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri/services -I /home/opam/.opam/4.14/lib/uutf -I src/dns/.dns.objs/byte -I src/dns_forward/.dns_forward.objs/byte -I src/dns_forward_lwt_unix/.dns_forward_lwt_unix.objs/byte -no-alias-deps -opaque -open Dune__exe -o src/dns_forward_test/.test_fake.eobjs/byte/dune__exe__Test.cmo -c -impl src/dns_forward_test/test.ml)
File "src/dns_forward_test/test.ml", line 1:
Error: Could not find the .cmi file for interface
src/dns_forward_test/test.mli.
```

I've no clue why this fails -- the first there's a dependency on `mirage-mtime.mock` in the dune file, where Mirage_mtime_set is part of; and the second error I don't understand why there's an attempt to find an interface file for test.ml. But I'm sure some build systems / dune person can solve these easily within seconds. It btw works fine on my local clone.

hannesm committed:
put both tests into a single executables stanza to avoid errors

hannesm commented:
I'm curious what the status is.it needs an upper bound on charrua (<2.1.0), but otherwise I guess it is fine for review & merge

@djs55 is this work mergeable, or is it too dangerous to do anything on vpnkit (due to risk of regressions)?

hannesm commented on vpnkit.opam:
```diff
@@ -27,44 +27,40 @@ depends: [
   "pcap-format" {>= "0.4.0"}
-  "cmdliner" {< "1.1.0"}
+  "cmdliner" {>= "1.1.0"}
   "charrua" {>= "1.3.0"}
```
```suggestion
  "charrua" {>= "1.3.0" & < "2.1.0"}
```

djs55 committed:
opam: add upper bound on charrua

Co-authored-by: Hannes Mehnert <hannes@mehnert.org>

Leonidas-from-XIV commented:
I think there's some DCOs missing but anything else that would block the PR being merged? I've seen that a lot of work went into updating the code and it would be a shame to have it go to waste. It would be nice to update to supported versions of the dependencies that are in the OPAM repo.
